src/receivers/HTTPReceiver.ts

```diff
@@ -392,8 +394,21 @@ export default class HTTPReceiver implements Receiver {
 
     // Handle custom routes
     if (Object.keys(this.routes).length) {
-      const match = this.routes[path] && this.routes[path][method] !== undefined;
-      if (match) { return this.routes[path][method](req, res); }
+      // Check if the request matches any of the custom routes
+      let pathMatch : string | boolean = false;
+      let params : ParamsDictionary = {};
+      Object.keys(this.routes).forEach((route) => {
+        const matchRegex = match(route, { decode: decodeURIComponent });
+        const tempMatch = matchRegex(path);
+        if (tempMatch) {
+          pathMatch = route;
+          params = tempMatch.params as ParamsDictionary;
+        }
```

Comment:
We might want to stop checking for matches once one is found to align with [the "first come first serve" routing precedence](https://stackoverflow.com/a/32604002) used by Express. As is, attempting to access `http://localhost:3000/crayon/blue` will return "There are no more blue crayons!".

```js
const app = new App({
  token: "xoxb-token-here",
  signingSecret: "example-signing-secret",
  customRoutes: [{
    path: '/crayon/blue',
    method: ['GET'],
    handler: (req, res) => {
      res.writeHead(200);
      res.end(`How cool! A true blue!`);
    },
  }, {
    path: '/crayon/:color',
    method: ['GET'],
    handler: (req, res) => {
      res.writeHead(200);
      res.end(`There are no more ${req.params.color} crayons!`);
    },
  }],
});
```

Reply:
And if not too much trouble, adding a test with multiple overlapping routes (to verify that this precedence is respected) would be wonderful!

Reply:
That is a very good point, I missed that. I have gone ahead and added a conditional return in the forEach and a test to make sure the first found route should be returned.

Reply:
I'm not sure that the conditional is enough to guarantee the ordering and have noticed it's still finicky with the above example 😕  Diving a bit deeper, it looks like [the `buildReceiverRoutes` function](https://github.com/slackapi/bolt-js/blob/00eed6f9235e8870297a80d12b22b5627a4317d2/src/receivers/custom-routes.ts#L17) is converting the `customRoutes` array to an object, causing these keys to have an undefined ordering.

Considering that the `ReceiverRoutes` type is only used in the context of building these routes, I think it's safe make changes to somehow hold these routes in an array to preserve the ordering while building the `ReceiverRoutes` objects.

Reply:
Thank you for being so patient with me @e-zim. Just so I understand, are you thinking it would be better to keep the `customRoutes` as an array instead of converting it to an object? Or just add `customRoutes` array to the `HTTPReceiver` class? Or am I missing the point entirely?

Reply:
No worries at all! To offer more context, the `customRoutes` array is converted into `ReceiverRoute`-like objects in the `buildReceiverRoutes` function and these objects are used in the `HTTPReceiver` and `SocketModeReceiver` classes. For this reason, the `customRoutes` array has to undergo a conversion and cannot be directly attached to the `routes` in these Receiver classes.

I would suggest changing the return type of the `buildReceiverRoutes` function to be a Map from the URL string to an array of a new `ReceiverRoute` type, with the `ReceiverRoute` type resembling [this object](https://github.com/slackapi/bolt-js/blob/00eed6f9235e8870297a80d12b22b5627a4317d2/src/receivers/custom-routes.ts#L13). This Map would preserve the order that entries are inserted, meaning the order of `customRoutes` can be maintained and relied upon when iterating over the `routes` in the `HTTPReceiver` and `SocketModeReceiver` classes.

Hoping this makes some sense, but apologies if this jumble of words is confusing. Happy to clarify more if needed!

Reply:
Thank you for the explanation, I believe I did it correctly. But I am happy to revise if this isn't what you were thinking.

Reply:
@e-zim bumping in case the last comment didn't notify you because I forgot to @ you

Reply:
Sorry for the delay! I have extended the example above to further test this and it appeared that my local `bolt-js` version was cached to a prior commit. Apologies for not catching this sooner! Removing my `node_modules` then reinstalling the packages with the latest changes seemed to fix things.

Thank you making the changes in 04ba9e0, but I'm no longer certain that these are required (through testing with the above example). If it is alright, would you be okay reverting this?

In searching for examples to test the supposed undefined object ordering, I found [the following from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys) stating that "the order of the array returned by `Object.keys()` is the same as that provided by a [`for...in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop", and that [the traversal order](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#description) of a `for...in` loop "is well-defined and consistent across implementations". This is confirmed to [hold true for V8](https://bugs.chromium.org/p/v8/issues/detail?id=3056#c8), so it seems like my confusion of caching got in the way of the Javascript standards 😖

I will add another comment about verifying the expected ordering behavior for this PR, but for future reference I don't believe testing an undefined object ordering can be done easily (if it's even relevant to your JS implementation). Sorry again for mistaking this behavior!